### PR TITLE
Implement server & client

### PR DESCRIPTION
**Pull request by ArdeleanuStefan**

### CODE REVIEW

File: ArdeleanuStefan_Task3/client.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      m_mq = std::make_unique<message_queue>(open_only, MESSAGE_QUEUE_NAME.c_str());
+      m_managed_smh = std::make_unique<managed_shared_memory>(open_only, SHARED_MEMORY_NAME.c_str());
+   }
+
+   void start()
+   {
+      MessageQueueRequest msg;
+
+      auto &mutex = *m_managed_smh->find_or_construct<interprocess_mutex>("mutex")();
+      auto &cond = *m_managed_smh->find_or_construct<interprocess_condition>("cond")();
```
Comment (maxlig79) [inline ArdeleanuStefan_Task3/client.cpp]: Maybe you should add in common.h consts for mutex and condition with better names than "mutex" and "condition".

File: ArdeleanuStefan_Task3/server.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+   }
+
+   void start()
+   {
+      unsigned int priority;
+      message_queue::size_type recvd_size;
+      MessageQueueRequest msg;
+
+      auto &mutex = *m_managed_smh->find_or_construct<interprocess_mutex>("mutex")();
+      auto &cond = *m_managed_smh->find_or_construct<interprocess_condition>("cond")();
```
Comment (maxlig79) [inline ArdeleanuStefan_Task3/server.cpp]: Maybe you should add in common.h consts for mutex and condition with better names than "mutex" and "condition".

File: ArdeleanuStefan_Task3/server.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+#include <boost/interprocess/allocators/allocator.hpp>
+#include <iostream>
+#include <boost/asio.hpp>
+#include <DynamicStringArray.hpp>
+#include "common.h"
+
+using namespace boost::interprocess;
+
+namespace Shared
+{
```
Comment (maxlig79) [inline ArdeleanuStefan_Task3/server.cpp]: The shared from the client is different than the shared from the server? If not maybe the namespace can be put in a common header.

File: ArdeleanuStefan_Task3/client.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            shared_memory_object::remove(SHARED_MEMORY_NAME.c_str());
+
+            break;
+         }
+
+         case CommandIds::ADD:
+         {
+            std::pair<Shared::String *, size_t> p = m_managed_smh->find<Shared::String>("ADD");
+            std::cout << p.first->c_str() << std::endl;
+
```
Comment (maxlig79) [inline ArdeleanuStefan_Task3/client.cpp]: You forgot to destroy the shared memory for ADD.

File: ArdeleanuStefan_Task3/server.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+   {
+      shared_memory_object::remove(SHARED_MEMORY_NAME.c_str());
+      message_queue::remove(MESSAGE_QUEUE_NAME.c_str());
+
+      m_mq = std::make_unique<message_queue>(create_only, MESSAGE_QUEUE_NAME.c_str(), MAX_MESSAGE_NUMBER, MAX_MESSAGE_SIZE);
+      m_managed_smh = std::make_unique<managed_shared_memory>(create_only, SHARED_MEMORY_NAME.c_str(), SHARED_MEMORY_SIZE);
+   }
+
+   void start()
+   {
```
Comment (maxlig79) [inline ArdeleanuStefan_Task3/server.cpp]: In this method there is a lot of common code, maybe you should consider to use inheritance and write only one external function.

**Review by maxlig79 — CHANGES_REQUESTED** [review]

File: ArdeleanuStefan_Task3/client.cpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+         scoped_lock<interprocess_mutex> lock(mutex);
+         cond.wait(lock);
+         
+         switch (command_pair.first)
+         {
+
+         case CommandIds::HELP:
+         {
+            std::pair<Shared::Vector *, size_t> vec = m_managed_smh->find<Shared::Vector>("HELP");
+
```
Comment (maxlig79) [inline ArdeleanuStefan_Task3/client.cpp]: Instead of "HELP" you should use the macro HELP_COMMAND.c_str()

**Review by maxlig79 — CHANGES_REQUESTED** [review]

**Review by maxlig79 — APPROVED** [review]

**Review by MariaCraciun1 — APPROVED** [review]

**Pull request merged**